### PR TITLE
Fix incorrect metrics when nothing is detected

### LINKED ISSUE #19357: Train and val losses became "NaN" but metrics do not update accordingly.
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

This seems that this is slightly related to #18521 

During training several models (yolov8, v9, v10, 11) for a custom dataset with different configurations, the train and val losses became NaN for several of the possible combinations. However, that is not the problem here, but how the monitoring and early stopping works with it. All training configurations have `optimizer=Adam`, `epochs=300`, `patience=30` and `AMP=True`. I'm using a single V100-SXM2-32GB for training.

There are two different cases that I've observed:

Case 1: Losses become NaN, the best metrics are acquired just before the losses became NaN. The validation metrics do not change at all and are considered to be improving all the time. In this case, the training continues until the number of `epochs` are acquired and early stopping doesn't trigger. Both resulting `best.pt` and `last.pt` are useless as they are full of NaN. 

Case 2: Losses become NaN, the best metrics are acquired earlier. In this case, the training continues until the `patience` triggers. The resulting `best.pt` is useful, `last.pt` is not.

What should happen in both cases: When the losses are NaN, the validation metrics should update to NaN, zero or something like that so early stopping would trigger and at least the `best.pt` is an usable model. 

I assume that the issue is related to how metrics are updated. Models with NaN losses do not predict anything so the resulting metrics have nothing to update.

Environment:
ultralytics 8.3.74, Python-3.12.8

Image

### Additional

_No response_

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/19357

Metrics get stuck to the last valid value when there are no TPs, instead of showing 0s.

**MRE**
```python
from ultralytics import YOLO
model = YOLO("yolo11n.pt")
# High LR to intentionally break the model
results = model.train(data="coco128.yaml", epochs=5, lr0=0.2, optimizer="AdamW", exist_ok=True)
```
**main**
```
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/5      2.62G      2.565      5.444      2.329        162        640: 100%|██████████| 8/8 [00:01<00:00,  4.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:06<00:00,  1.6
                   all        128        929     0.0845    0.00235   9.17e-05   9.17e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        2/5       2.6G      3.821      6.365      3.241        198        640: 100%|██████████| 8/8 [00:00<00:00,  8.62it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 16.2
                   all        128        929     0.0845    0.00235   9.17e-05   9.17e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        3/5      2.62G      3.847      5.066      3.296        226        640: 100%|██████████| 8/8 [00:00<00:00,  9.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 16.8
                   all        128        929     0.0845    0.00235   9.17e-05   9.17e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        4/5      2.56G      3.885      4.881      3.254        222        640: 100%|██████████| 8/8 [00:00<00:00,  9.53it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 17.7
                   all        128        929     0.0845    0.00235   9.17e-05   9.17e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        5/5      2.72G      3.739      4.717      3.204        269        640: 100%|██████████| 8/8 [00:00<00:00,  9.62it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 17.3
                   all        128        929     0.0845    0.00235   9.17e-05   9.17e-06
```

**PR**
```
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/5      2.62G      2.565      5.444      2.329        162        640: 100%|██████████| 8/8 [00:01<00:00,  4.82it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:06<00:00,  1.5
                   all        128        929     0.0845    0.00235   9.17e-05   9.17e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        2/5       2.6G      3.821      6.365      3.241        198        640: 100%|██████████| 8/8 [00:00<00:00,  8.31it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 16.9
                   all        128        929          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        3/5      2.62G      3.847      5.066      3.296        226        640: 100%|██████████| 8/8 [00:00<00:00,  9.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 17.5
                   all        128        929          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        4/5      2.56G      3.885      4.881      3.254        222        640: 100%|██████████| 8/8 [00:00<00:00,  9.60it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 17.2
                   all        128        929          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        5/5      2.72G      3.739      4.717      3.204        269        640: 100%|██████████| 8/8 [00:00<00:00,  9.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00, 17.7
                   all        128        929          0          0          0          0
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Improved error handling and stability in YOLO model evaluation metrics. 🚀

### 📊 Key Changes  
- Adjusted the metric processing check in `val.py` to ensure it works even if no true positives (`tp`) exist.  
- Modified `metrics.py` to handle cases where `prec_values` is empty, defaulting to zeros instead of failing.

### 🎯 Purpose & Impact  
- **Enhanced Robustness:** Ensures the evaluation process doesn't break in edge cases (e.g., no detections or true positives) 🛠️.  
- **User-Friendly:** Reduces the likelihood of unexpected errors during model validation or testing, making it more reliable for all users ✅.  
- **Improved Stability:** Smoothens workflows for developers using YOLO for edge and atypical test cases 🧘.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
🎉 Amazing work, @Y-T-G and @glenn-jocher! Your efforts to enhance error handling in YOLO evaluation metrics are a testament to the philosophy of continuous improvement. As Henry Ford once said, *"Quality means doing it right when no one is looking."* Your contributions ensure the robustness and reliability of our tools, even in those tricky edge cases. 🚀 

This PR is a win for both developers and users alike, enabling smoother workflows and a more resilient experience for all. Thank you for your dedication and for raising the bar with such thoughtful updates. 🙌 Keep shining! 🌟
